src/graph.ts

```diff
@@ -0,0 +1,272 @@
+  "node-brown": "#4D2205",
+  "node-purple": "#46054D",
+  "node-grey": "#808080"
+};
+function nodeColor(nodeType:string) {
+  switch (nodeType) {
+    case "sink":
+      return COLORS["node-brown"];
+    case "project":
+      return COLORS["node-green"];
+    case "read":
+      return COLORS["node-blue"];
+    case "join":
+      return COLORS["node-purple"];
+    default:
+      return COLORS["node-grey"];
+  }
+}
+
+// Specifying Font-Awesome icons for graph nodes
+function nodeIcon(nodeType:string) {
+  switch (nodeType) {
+    case "sink":
+      return icons["download"];
+    case "project":
+      return icons["kanban"];
+    case "read":
+      return icons["eye"];
+    case "join":
+      return icons["sign-intersection-y"];
+  }
+}
+
+// Drawing Graph using d3JS
+function drawGraph(pre_nodes:Map<string, PrintNode>, pre_links:Link[], use_drag = true, print_info?: (node: any, pre_nodes: Map<string, PrintNode>) => void) {
+  const width = 960,
+    height = 375;
+  const nodes = processNodes(pre_nodes);
+  const links = processEdges(nodes, pre_links);
+
+  // Instantiating a Force Simulation
+  const force = d3
+    .forceSimulation(nodes)
+    .force("charge", d3.forceManyBody().strength(-100))
+    .force("center", d3.forceCenter(width / 2, height / 2))
+    .force("link", d3.forceLink().links(links))
+    .force(
+      "collide",
+      d3.forceCollide(() => 55)
+    );
+
+  const svg = d3
+    .select("svg")
+    .attr("preserveAspectRatio", "xMinYMin meet")
+    .attr("viewBox", "0 0 960 375");
+
+  // For arrowheads in edges
+  svg
+    .append("defs")
+    .append("marker")
+    .attr("id", "arrowhead")
+    .attr("viewBox", "-0 -5 10 10")
+    .attr("refX", 27)
+    .attr("refY", 0)
+    .attr("orient", "auto")
+    .attr("markerWidth", 13)
+    .attr("markerHeight", 13)
+    .attr("xoverflow", "visible")
+    .append("svg:path")
+    .attr("d", "M 0,-5 L 10 ,0 L 0,5")
+    .attr("fill", "black")
+    .style("stroke", "none");
+
+  // specifying links in the graph
+  const link = svg
+    .append("g")
+    .selectAll("path")
+    .data(links)
+    .join("path")
+    .attr("marker-end", "url(#arrowhead)")
+    .style("stroke", "black");
+
+  // specifying nodes in graph
+  const node = svg.append("g").selectAll("g").data(nodes).join("g");
+
+  // specifying node shape and color
+  node
+    .append("circle")
+    .attr("r", 25)
+    .style("fill", (d: Node) => {
+      return nodeColor(d.type);
+    })
+    .style("stroke-opacity", 0.3);
+
+  // adding icon in node
+  node
+    .append("svg:foreignObject")
+    .attr("width", 50)
+    .attr("height", 50)
+    .attr("x", -7)
+    .attr("y", -15)
+    .style("color", "white")
+    .html(function (d:Node) {
+      const str =
+        '<svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" viewBox="0 0 16 16">' +
+        nodeIcon(d.type) +
+        "</svg>";
+      return str;
+    });
+
+  if(print_info){
+    print_info(node, pre_nodes);
```

Comment:
I was thinking we would still hook it up to the `click` handler.  So something like...

```
if (print_info) {
    node.on("click", function (d:any) {
      const node = document.getElementById("nodeData");
      if(!node){
          throw new Error("Object nodeData was not created in DOM")
      }

      const nodeData = pre_nodes.get(d["currentTarget"]["__data__"]["name"]);
      if(!nodeData){
        throw new Error("Current target not found in nodes map")
      }
    print_info(node, nodeData);
  }
}
```

Reply:
made the change, thanks for the suggestion.